lib/ansible/modules/cloud/azure/azure_vm_metadata_facts.py

```diff
@@ -0,0 +1,369 @@
+            #
+            newkey = re.sub(':|-', '_', key).lower()
+
+            fields[newkey] = fields.pop(key)
+            #
+            # if list, roll through the list and parse what is inside
+            #
+            if isinstance(val, list):
+                templist = []
+                for l in val:
+                    templist.append(self._keyfix(l))
+                fields[newkey] = templist
+            #
+            # if dictionary, recurse through
+            #
+            elif isinstance(val, dict):
+                fields[newkey] = self._keyfix(val)
+            else:
+                fields[newkey] = val
+
+        return fields
+
+
+    #
+    # Designed to create template-friendly variable names.
+    # While keyfix just returns a dictionary, this function
+    # seperates every key+value in to a variable.
+    #
+    def _json_to_facts(self, fields, prefix_key=""):
+
+        if (not prefix_key):
+            prefix_key = (self._prefix % "_")
+
+        for key, val in fields.items():
+
+            #
+            # Change ':'' and '-' to '_' to ensure valid template variable names
+            #
+            newkey = prefix_key + re.sub(':|-', '_', key).lower()
+            #
+            # if list, roll through the list and parse what is inside
+            #
+            if isinstance(val, list):
+                for l in val:
+                    self._json_to_facts(l, str(newkey) + "_")
+            #
+            # if dictionary, recurse through
+            #
+            elif isinstance(val, dict):
+                self._json_to_facts(val, str(newkey) + "_")
+            else:
+                self._data[newkey] = val
+
+
+    #
+    # Take Metadata results and parse JSON
+    #
+    def fetch(self, uri):
+
+        raw_results = self._fetch(uri)
+
+        if not raw_results:
+            return
+        #
+        # There should not be any newlines.
+        # Handle the newlines if they exist.
+        #
+        raw_result = raw_results.split('\n')
+
+        for result in raw_result:
+
+            try:
+                #
+                # Results should be a JSON string.
+                # Load the JSON in to a dictionary.
+                #
+                self._data[self._prefix % ""] = json.loads(result)
+                for (key, val) in self._data.items():
+                    newkey = re.sub(':|-', '_', key).lower()
+
+                    #
+                    # Change the key name to the ansible-friendly version
+                    #
+                    self._data[newkey] = self._data.pop(key)
+                    #
+                    # 
+                    if isinstance(val, dict):
+                        self._data[newkey] = self._keyfix(val)
+                self._json_to_facts(self._data[self._prefix % ""])
+
+            except Exception as e:
+                self.module.warn("Exception: " + str(e))
+
+
+    def run(self):
```

Comment:
Replace this method with `fetch`, with return value.

Reply:
@yuwzho , what do you want to replace with `fetch` ?
`self.module.warn("Exception: " + str(e))` ?
`def run(self):` ?

`def run(self):` already leverages `fetch`
```
    def run(self):
        self.fetch(self.uri_meta)
        return self._data
```

I don't understand how `self.module.warn("Exception: " + str(e))`  should be modified to use `fetch`.

Please post a code example of what you are asking for so I can better understand.

Reply:
IMO, there is no sense to add an additional method call for `self.fetch`, putting the logic to `self.run` is enough. Then the fetch should maintain a local dict rather than changing a global variable frequently